File: storm-core/test/clj/org/apache/storm/security/auth/nimbus_auth_test.clj
;; Licensed to the Apache Software Foundation (ASF) under one
;; or more contributor license agreements.  See the NOTICE file
;; distributed with this work for additional information
;; regarding copyright ownership.  The ASF licenses this file
;; to you under the Apache License, Version 2.0 (the
;; "License"); you may not use this file except in compliance
;; with the License.  You may obtain a copy of the License at
;;
;; http://www.apache.org/licenses/LICENSE-2.0
;;
;; Unless required by applicable law or agreed to in writing, software
;; distributed under the License is distributed on an "AS IS" BASIS,
;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
;; See the License for the specific language governing permissions and
;; limitations under the License.
(ns org.apache.storm.security.auth.nimbus-auth-test
  (:use [clojure test])
  (:require [org.apache.storm.security.auth [auth-test :refer [nimbus-timeout]]])
  (:import [java.nio ByteBuffer])
  (:import [java.util Optional])
  (:import [org.apache.storm LocalCluster$Builder DaemonConfig Config])
  (:import [org.apache.storm.blobstore BlobStore])
  (:import [org.apache.storm.utils NimbusClient])
  (:import [org.apache.storm.generated NotAliveException StormBase])
  (:import [org.apache.storm.security.auth AuthUtils ThriftServer ThriftClient
                                         ReqContext ThriftConnectionType])
  (:import [org.apache.storm.generated Nimbus Nimbus$Client Nimbus$Processor
            AuthorizationException SubmitOptions TopologyInitialStatus KillOptions])
  (:import [org.apache.storm.utils ConfigUtils Utils])
  (:import [org.apache.storm.cluster IStormClusterState])
  (:import [org.mockito Mockito Matchers])
  (:use [org.apache.storm util config daemon-config log])
  (:require [conjure.core])
  (:use [conjure core]))

(defn to-conf [nimbus-port login-cfg aznClass transportPluginClass]
  (let [conf {NIMBUS-AUTHORIZER aznClass
              NIMBUS-THRIFT-PORT nimbus-port
              STORM-THRIFT-TRANSPORT-PLUGIN transportPluginClass }
         conf (if login-cfg (merge conf {"java.security.auth.login.config" login-cfg}) conf)]
    conf))

(defmacro with-test-cluster [[cluster-sym & args] & body]
  `(let [conf# (to-conf ~@args)]
     (with-open [~cluster-sym (.build (doto (LocalCluster$Builder. )
                      (.withNimbusDaemon)
                      (.withDaemonConf conf#)
                      (.withSupervisors 0)
                      (.withPortsPerSupervisor 0)))]
       ~@body)))

(deftest Simple-authentication-test
  (with-test-cluster [cluster 0 nil nil "org.apache.storm.security.auth.SimpleTransportPlugin"]
    (let [storm-conf (merge (clojurify-structure (ConfigUtils/readStormConfig))
                            {STORM-THRIFT-TRANSPORT-PLUGIN "org.apache.storm.security.auth.SimpleTransportPlugin"
                             STORM-NIMBUS-RETRY-TIMES 0})
          client (NimbusClient. storm-conf "localhost" (.getThriftServerPort cluster) nimbus-timeout)
          nimbus_client (.getClient client)]
      (testing "(Positive authorization) Simple protocol w/o authentication/authorization enforcement"
               (is (thrown-cause? NotAliveException
                            (.activate nimbus_client "topo-name"))))
      (.close client))))

(deftest test-noop-authorization-w-simple-transport
  (let [cluster-state (Mockito/mock IStormClusterState)
        blob-store (Mockito/mock BlobStore)
        topo-name "topo-name"]
    (.thenReturn (Mockito/when (.getTopoId cluster-state topo-name)) (Optional/empty))
    (with-open [cluster (.build
                          (doto (LocalCluster$Builder.)
                            (.withClusterState cluster-state)
                            (.withBlobStore blob-store)
                            (.withNimbusDaemon)
                            (.withDaemonConf
                               {NIMBUS-AUTHORIZER "org.apache.storm.security.auth.authorizer.NoopAuthorizer"
                                NIMBUS-THRIFT-PORT 0
                                STORM-THRIFT-TRANSPORT-PLUGIN "org.apache.storm.security.auth.SimpleTransportPlugin"})))]
      (let [storm-conf (merge (clojurify-structure (ConfigUtils/readStormConfig))
                               {STORM-THRIFT-TRANSPORT-PLUGIN "org.apache.storm.security.auth.SimpleTransportPlugin"
                                STORM-NIMBUS-RETRY-TIMES 0})
            client (NimbusClient. storm-conf "localhost" (.getThriftServerPort cluster) nimbus-timeout)
            nimbus_client (.getClient client)]
        (testing "(Positive authorization) Authorization plugin should accept client request"
                 (is (thrown-cause? NotAliveException
                              (.activate nimbus_client topo-name))))
        (.close client)))))

(deftest test-deny-authorization-w-simple-transport
  (let [cluster-state (Mockito/mock IStormClusterState)
        blob-store (Mockito/mock BlobStore)
        topo-name "topo-name"
        topo-id "topo-name-1"]
    (.thenReturn (Mockito/when (.getTopoId cluster-state topo-name)) (Optional/of topo-id))
    (.thenReturn (Mockito/when (.readTopologyConf blob-store (Mockito/any String) (Mockito/anyObject))) {})
    (with-open [cluster (.build
                          (doto (LocalCluster$Builder.)
                            (.withClusterState cluster-state)
                            (.withBlobStore blob-store)
                            (.withNimbusDaemon)
                            (.withDaemonConf
                               {NIMBUS-AUTHORIZER "org.apache.storm.security.auth.authorizer.DenyAuthorizer"
                                NIMBUS-THRIFT-PORT 0
                                STORM-THRIFT-TRANSPORT-PLUGIN "org.apache.storm.security.auth.SimpleTransportPlugin"})))]
      (let [storm-conf (merge (clojurify-structure (ConfigUtils/readStormConfig))
                              {STORM-THRIFT-TRANSPORT-PLUGIN   "org.apache.storm.security.auth.SimpleTransportPlugin"
                               Config/NIMBUS_THRIFT_PORT (.getThriftServerPort cluster)
                               STORM-NIMBUS-RETRY-TIMES        0})
            client (NimbusClient. storm-conf "localhost" (.getThriftServerPort cluster) nimbus-timeout)
            nimbus_client (.getClient client)
            topologyInitialStatus (TopologyInitialStatus/findByValue 2)
            submitOptions (SubmitOptions. topologyInitialStatus)]
        (is (thrown-cause? AuthorizationException (.submitTopology nimbus_client topo-name nil nil nil)))
        (is (thrown-cause? AuthorizationException (.submitTopologyWithOpts nimbus_client topo-name nil nil nil submitOptions)))
        (is (thrown-cause? AuthorizationException (.beginFileUpload nimbus_client)))

        (is (thrown-cause? AuthorizationException (.uploadChunk nimbus_client nil nil)))
        (is (thrown-cause? AuthorizationException (.finishFileUpload nimbus_client nil)))
        (is (thrown-cause? AuthorizationException (.beginFileDownload nimbus_client nil)))
        (is (thrown-cause? AuthorizationException (.downloadChunk nimbus_client nil)))
        (is (thrown-cause? AuthorizationException (.getNimbusConf nimbus_client)))
        (is (thrown-cause? AuthorizationException (.getClusterInfo nimbus_client)))
        (is (thrown-cause? AuthorizationException (.killTopology nimbus_client topo-name)))
        (is (thrown-cause? AuthorizationException (.killTopologyWithOpts nimbus_client topo-name (KillOptions.))))
        (is (thrown-cause? AuthorizationException (.activate nimbus_client topo-name)))
        (is (thrown-cause? AuthorizationException (.deactivate nimbus_client topo-name)))
        (is (thrown-cause? AuthorizationException (.rebalance nimbus_client topo-name nil)))
        (is (thrown-cause? AuthorizationException (.getTopologyConf nimbus_client topo-id)))
        (is (thrown-cause? AuthorizationException (.getTopology nimbus_client topo-id)))
        (is (thrown-cause? AuthorizationException (.getUserTopology nimbus_client topo-id)))
        (is (thrown-cause? AuthorizationException (.getTopologyInfo nimbus_client topo-id)))
        (.close client)))))

(deftest test-noop-authorization-w-sasl-digest
  (with-test-cluster [cluster 0
                "test/clj/org/apache/storm/security/auth/jaas_digest.conf"
                "org.apache.storm.security.auth.authorizer.NoopAuthorizer"
                "org.apache.storm.security.auth.digest.DigestSaslTransportPlugin"]
    (let [storm-conf (merge (clojurify-structure (ConfigUtils/readStormConfig))
                            {STORM-THRIFT-TRANSPORT-PLUGIN "org.apache.storm.security.auth.digest.DigestSaslTransportPlugin"
                             "java.security.auth.login.config" "test/clj/org/apache/storm/security/auth/jaas_digest.conf"
                             Config/NIMBUS_THRIFT_PORT (.getThriftServerPort cluster)
                             STORM-NIMBUS-RETRY-TIMES 0})
          client (NimbusClient. storm-conf "localhost" (.getThriftServerPort cluster) nimbus-timeout)
          nimbus_client (.getClient client)]
      (testing "(Positive authorization) Authorization plugin should accept client request"
               (is (thrown-cause? NotAliveException
                            (.activate nimbus_client "topo-name"))))
      (.close client))))

(deftest test-deny-authorization-w-sasl-digest
  (let [cluster-state (Mockito/mock IStormClusterState)
        blob-store (Mockito/mock BlobStore)
        topo-name "topo-name"
        topo-id "topo-name-1"]
    (.thenReturn (Mockito/when (.getTopoId cluster-state topo-name)) (Optional/of topo-id))
    (.thenReturn (Mockito/when (.readTopologyConf blob-store (Mockito/any String) (Mockito/anyObject))) {})
    (with-open [cluster (.build
                          (doto (LocalCluster$Builder.)
                            (.withClusterState cluster-state)
                            (.withBlobStore blob-store)
                            (.withNimbusDaemon)
                            (.withDaemonConf
                               {NIMBUS-AUTHORIZER "org.apache.storm.security.auth.authorizer.DenyAuthorizer"
                                NIMBUS-THRIFT-PORT 0
                                "java.security.auth.login.config" "test/clj/org/apache/storm/security/auth/jaas_digest.conf"
                                STORM-THRIFT-TRANSPORT-PLUGIN "org.apache.storm.security.auth.digest.DigestSaslTransportPlugin"})))]
      (let [storm-conf (merge (clojurify-structure (ConfigUtils/readStormConfig))
                               {STORM-THRIFT-TRANSPORT-PLUGIN "org.apache.storm.security.auth.digest.DigestSaslTransportPlugin"
                               "java.security.auth.login.config" "test/clj/org/apache/storm/security/auth/jaas_digest.conf"
                               Config/NIMBUS_THRIFT_PORT (.getThriftServerPort cluster)
                               STORM-NIMBUS-RETRY-TIMES 0})
            client (NimbusClient. storm-conf "localhost" (.getThriftServerPort cluster) nimbus-timeout)
            nimbus_client (.getClient client)
            topologyInitialStatus (TopologyInitialStatus/findByValue 2)
            submitOptions (SubmitOptions. topologyInitialStatus)]
        (is (thrown-cause? AuthorizationException (.submitTopology nimbus_client topo-name nil nil nil)))
        (is (thrown-cause? AuthorizationException (.submitTopologyWithOpts nimbus_client topo-name nil nil nil submitOptions)))
        (is (thrown-cause? AuthorizationException (.beginFileUpload nimbus_client)))

        (is (thrown-cause? AuthorizationException (.uploadChunk nimbus_client nil nil)))
        (is (thrown-cause? AuthorizationException (.finishFileUpload nimbus_client nil)))
        (is (thrown-cause? AuthorizationException (.beginFileDownload nimbus_client nil)))
        (is (thrown-cause? AuthorizationException (.downloadChunk nimbus_client nil)))
        (is (thrown-cause? AuthorizationException (.getNimbusConf nimbus_client)))
        (is (thrown-cause? AuthorizationException (.getClusterInfo nimbus_client)))
        (is (thrown-cause? AuthorizationException (.killTopology nimbus_client topo-name)))
        (is (thrown-cause? AuthorizationException (.killTopologyWithOpts nimbus_client topo-name (KillOptions.))))
        (is (thrown-cause? AuthorizationException (.activate nimbus_client topo-name)))
        (is (thrown-cause? AuthorizationException (.deactivate nimbus_client topo-name)))
        (is (thrown-cause? AuthorizationException (.rebalance nimbus_client topo-name nil)))
        (is (thrown-cause? AuthorizationException (.getTopologyConf nimbus_client topo-id)))
        (is (thrown-cause? AuthorizationException (.getTopology nimbus_client topo-id)))
        (is (thrown-cause? AuthorizationException (.getUserTopology nimbus_client topo-id)))
        (is (thrown-cause? AuthorizationException (.getTopologyInfo nimbus_client topo-id)))
        (.close client)))))

File: storm-server/src/test/java/org/apache/storm/security/auth/NimbusAuthTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.storm.security.auth;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import org.apache.storm.Config;
import org.apache.storm.DaemonConfig;
import org.apache.storm.LocalCluster;
import org.apache.storm.blobstore.BlobStore;
import org.apache.storm.cluster.IStormClusterState;
import org.apache.storm.daemon.nimbus.TopoCache;
import org.apache.storm.generated.AuthorizationException;
import org.apache.storm.generated.KillOptions;
import org.apache.storm.generated.Nimbus;
import org.apache.storm.generated.NotAliveException;
import org.apache.storm.generated.SubmitOptions;
import org.apache.storm.generated.TopologyInitialStatus;
import org.apache.storm.utils.ConfigUtils;
import org.apache.storm.utils.NimbusClient;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentMatchers;
import org.mockito.Mockito;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for Nimbus authentication and authorization with various transport plugins.
 *
 * Ported from storm-core/test/clj/org/apache/storm/security/auth/nimbus_auth_test.clj
 */
public class NimbusAuthTest {

    private static final int NIMBUS_TIMEOUT = 3000;
    private static final String JAAS_CONF = jaasConfPath();

    private static String jaasConfPath() {
        java.net.URL url = NimbusAuthTest.class.getResource("/org/apache/storm/security/auth/jaas_digest.conf");
        if (url == null) {
            throw new RuntimeException("jaas_digest.conf not found on classpath");
        }
        return url.getPath();
    }

    /**
     * Checks if the given exception or any of its causes is an instance of the expected class.
     */
    private static boolean hasCause(Throwable t, Class<? extends Throwable> expected) {
        Throwable current = t;
        while (current != null) {
            if (expected.isInstance(current)) {
                return true;
            }
            current = current.getCause();
        }
        return false;
    }

    @FunctionalInterface
    private interface ThrowingRunnable {
        void run() throws Exception;
    }

    private static void assertThrowsCause(Class<? extends Throwable> expectedCause,
                                           ThrowingRunnable action) {
        try {
            action.run();
            fail("Expected exception with cause " + expectedCause.getSimpleName() + " but no exception was thrown");
        } catch (Exception e) {
            assertTrue(hasCause(e, expectedCause),
                "Expected cause " + expectedCause.getSimpleName() + " but got: " + e);
        }
    }

    @Test
    public void testSimpleAuthentication() throws Exception {
        Map<String, Object> clusterConf = new HashMap<>();
        clusterConf.put(DaemonConfig.NIMBUS_AUTHORIZER, null);
        clusterConf.put(DaemonConfig.SUPERVISOR_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.NoopAuthorizer");
        clusterConf.put(Config.NIMBUS_THRIFT_PORT, 0);
        clusterConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
            "org.apache.storm.security.auth.SimpleTransportPlugin");

        try (LocalCluster cluster = new LocalCluster.Builder()
                .withNimbusDaemon()
                .withDaemonConf(clusterConf)
                .withSupervisors(0)
                .withPortsPerSupervisor(0)
                .build()) {

            Map<String, Object> clientConf = new HashMap<>(ConfigUtils.readStormConfig());
            clientConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
                "org.apache.storm.security.auth.SimpleTransportPlugin");
            clientConf.put(Config.STORM_NIMBUS_RETRY_TIMES, 0);

            try (NimbusClient client = new NimbusClient(clientConf, "localhost",
                    cluster.getThriftServerPort(), NIMBUS_TIMEOUT)) {
                Nimbus.Iface nimbusClient = client.getClient();
                assertThrowsCause(NotAliveException.class, () -> nimbusClient.activate("topo-name"));
            }
        }
    }

    @Test
    public void testNoopAuthorizationWithSimpleTransport() throws Exception {
        IStormClusterState clusterState = Mockito.mock(IStormClusterState.class);
        BlobStore blobStore = Mockito.mock(BlobStore.class);
        TopoCache tc = Mockito.mock(TopoCache.class);
        String topoName = "topo-name";

        Mockito.when(clusterState.getTopoId(topoName)).thenReturn(Optional.empty());

        Map<String, Object> clusterConf = new HashMap<>();
        clusterConf.put(DaemonConfig.NIMBUS_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.NoopAuthorizer");
        clusterConf.put(DaemonConfig.SUPERVISOR_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.NoopAuthorizer");
        clusterConf.put(Config.NIMBUS_THRIFT_PORT, 0);
        clusterConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
            "org.apache.storm.security.auth.SimpleTransportPlugin");

        try (LocalCluster cluster = new LocalCluster.Builder()
                .withClusterState(clusterState)
                .withBlobStore(blobStore)
                .withTopoCache(tc)
                .withNimbusDaemon()
                .withDaemonConf(clusterConf)
                .build()) {

            Map<String, Object> clientConf = new HashMap<>(ConfigUtils.readStormConfig());
            clientConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
                "org.apache.storm.security.auth.SimpleTransportPlugin");
            clientConf.put(Config.STORM_NIMBUS_RETRY_TIMES, 0);

            try (NimbusClient client = new NimbusClient(clientConf, "localhost",
                    cluster.getThriftServerPort(), NIMBUS_TIMEOUT)) {
                Nimbus.Iface nimbusClient = client.getClient();
                assertThrowsCause(NotAliveException.class, () -> nimbusClient.activate(topoName));
            }
        }
    }

    @Test
    public void testDenyAuthorizationWithSimpleTransport() throws Exception {
        IStormClusterState clusterState = Mockito.mock(IStormClusterState.class);
        BlobStore blobStore = Mockito.mock(BlobStore.class);
        TopoCache tc = Mockito.mock(TopoCache.class);
        String topoName = "topo-name";
        String topoId = "topo-name-1";

        Mockito.when(clusterState.getTopoId(topoName)).thenReturn(Optional.of(topoId));
        Mockito.when(tc.readTopoConf(Mockito.any(String.class), ArgumentMatchers.any()))
            .thenReturn(Map.of());

        Map<String, Object> clusterConf = new HashMap<>();
        clusterConf.put(DaemonConfig.NIMBUS_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.DenyAuthorizer");
        clusterConf.put(DaemonConfig.SUPERVISOR_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.DenyAuthorizer");
        clusterConf.put(Config.NIMBUS_THRIFT_PORT, 0);
        clusterConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
            "org.apache.storm.security.auth.SimpleTransportPlugin");

        try (LocalCluster cluster = new LocalCluster.Builder()
                .withClusterState(clusterState)
                .withBlobStore(blobStore)
                .withTopoCache(tc)
                .withNimbusDaemon()
                .withDaemonConf(clusterConf)
                .build()) {

            Map<String, Object> clientConf = new HashMap<>(ConfigUtils.readStormConfig());
            clientConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
                "org.apache.storm.security.auth.SimpleTransportPlugin");
            clientConf.put(Config.NIMBUS_THRIFT_PORT, cluster.getThriftServerPort());
            clientConf.put(Config.STORM_NIMBUS_RETRY_TIMES, 0);

            try (NimbusClient client = new NimbusClient(clientConf, "localhost",
                    cluster.getThriftServerPort(), NIMBUS_TIMEOUT)) {
                Nimbus.Iface nimbusClient = client.getClient();
                SubmitOptions submitOptions = new SubmitOptions(TopologyInitialStatus.findByValue(2));

                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.submitTopology(topoName, null, null, null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.submitTopologyWithOpts(topoName, null, null, null, submitOptions));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.beginFileUpload());
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.uploadChunk(null, null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.finishFileUpload(null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.downloadChunk(null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getNimbusConf());
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getClusterInfo());
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.killTopology(topoName));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.killTopologyWithOpts(topoName, new KillOptions()));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.activate(topoName));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.deactivate(topoName));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.rebalance(topoName, null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getTopologyConf(topoId));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getTopology(topoId));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getUserTopology(topoId));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getTopologyInfo(topoId));
            }
        }
    }

    @Test
    public void testNoopAuthorizationWithSaslDigest() throws Exception {
        Map<String, Object> clusterConf = new HashMap<>();
        clusterConf.put(DaemonConfig.NIMBUS_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.NoopAuthorizer");
        clusterConf.put(DaemonConfig.SUPERVISOR_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.NoopAuthorizer");
        clusterConf.put(Config.NIMBUS_THRIFT_PORT, 0);
        clusterConf.put("java.security.auth.login.config",
            JAAS_CONF);
        clusterConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
            "org.apache.storm.security.auth.digest.DigestSaslTransportPlugin");

        try (LocalCluster cluster = new LocalCluster.Builder()
                .withNimbusDaemon()
                .withDaemonConf(clusterConf)
                .withSupervisors(0)
                .withPortsPerSupervisor(0)
                .build()) {

            Map<String, Object> clientConf = new HashMap<>(ConfigUtils.readStormConfig());
            clientConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
                "org.apache.storm.security.auth.digest.DigestSaslTransportPlugin");
            clientConf.put("java.security.auth.login.config",
                JAAS_CONF);
            clientConf.put(Config.NIMBUS_THRIFT_PORT, cluster.getThriftServerPort());
            clientConf.put(Config.STORM_NIMBUS_RETRY_TIMES, 0);

            try (NimbusClient client = new NimbusClient(clientConf, "localhost",
                    cluster.getThriftServerPort(), NIMBUS_TIMEOUT)) {
                Nimbus.Iface nimbusClient = client.getClient();
                assertThrowsCause(NotAliveException.class, () -> nimbusClient.activate("topo-name"));
            }
        }
    }

    @Test
    public void testDenyAuthorizationWithSaslDigest() throws Exception {
        IStormClusterState clusterState = Mockito.mock(IStormClusterState.class);
        BlobStore blobStore = Mockito.mock(BlobStore.class);
        TopoCache tc = Mockito.mock(TopoCache.class);
        String topoName = "topo-name";
        String topoId = "topo-name-1";

        Mockito.when(clusterState.getTopoId(topoName)).thenReturn(Optional.of(topoId));
        Mockito.when(tc.readTopoConf(Mockito.any(String.class), ArgumentMatchers.any()))
            .thenReturn(Map.of());

        Map<String, Object> clusterConf = new HashMap<>();
        clusterConf.put(DaemonConfig.NIMBUS_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.DenyAuthorizer");
        clusterConf.put(DaemonConfig.SUPERVISOR_AUTHORIZER,
            "org.apache.storm.security.auth.authorizer.DenyAuthorizer");
        clusterConf.put(Config.NIMBUS_THRIFT_PORT, 0);
        clusterConf.put("java.security.auth.login.config",
            JAAS_CONF);
        clusterConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
            "org.apache.storm.security.auth.digest.DigestSaslTransportPlugin");

        try (LocalCluster cluster = new LocalCluster.Builder()
                .withClusterState(clusterState)
                .withBlobStore(blobStore)
                .withTopoCache(tc)
                .withNimbusDaemon()
                .withDaemonConf(clusterConf)
                .build()) {

            Map<String, Object> clientConf = new HashMap<>(ConfigUtils.readStormConfig());
            clientConf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN,
                "org.apache.storm.security.auth.digest.DigestSaslTransportPlugin");
            clientConf.put("java.security.auth.login.config",
                JAAS_CONF);
            clientConf.put(Config.NIMBUS_THRIFT_PORT, cluster.getThriftServerPort());
            clientConf.put(Config.STORM_NIMBUS_RETRY_TIMES, 0);

            try (NimbusClient client = new NimbusClient(clientConf, "localhost",
                    cluster.getThriftServerPort(), NIMBUS_TIMEOUT)) {
                Nimbus.Iface nimbusClient = client.getClient();
                SubmitOptions submitOptions = new SubmitOptions(TopologyInitialStatus.findByValue(2));

                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.submitTopology(topoName, null, null, null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.submitTopologyWithOpts(topoName, null, null, null, submitOptions));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.beginFileUpload());
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.uploadChunk(null, null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.finishFileUpload(null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.downloadChunk(null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getNimbusConf());
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getClusterInfo());
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.killTopology(topoName));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.killTopologyWithOpts(topoName, new KillOptions()));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.activate(topoName));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.deactivate(topoName));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.rebalance(topoName, null));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getTopologyConf(topoId));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getTopology(topoId));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getUserTopology(topoId));
                assertThrowsCause(AuthorizationException.class,
                    () -> nimbusClient.getTopologyInfo(topoId));
            }
        }
    }
}
